### PR TITLE
limiting typeahead

### PR DESCRIPTION
**Pull request by LindsayYoung**
This will decrease the chances of offensive names popping up in typeahead.

This is not perfect because there are some issues with dimcandproperties. there are candidates that should show up that don't.

fix 18F/openFEC-web-app#804

### CODE REVIEW

File: webservices/args.py [hunk 1]
```diff
@@ -172,7 +172,7 @@ def one_of(value, options):
     'office': Arg(str, multiple=True, enum=['', 'H', 'S', 'P'], description='Governmental office candidate runs for: House, Senate or President.'),
     'state': IString(multiple=True, description='U.S. State candidate or territory where a candidate runs for office.'),
     'party': IString(multiple=True, description='Three letter code for the party under which a candidate ran for office'),
-    'year': Arg(str, dest='election_year', description='See records pertaining to a particular election year.'),
+    'year': Arg(str, dest='election_year', description='See records pertaining to a particular election year. The list of election years is based on a candidate filing a F2 for that year.'),
```
Comment (jmcarp) [inline webservices/args.py]: "...filing an F2..."?

**Comment by jmcarp** [comment]
@LindsayYoung can you say more about the candidates that should show up but don't? Are there specific categories of candidates who won't show up, and are there many of them? Are there data errors that should be fixed related to this?

**Comment by LindsayYoung** [comment]
There are two places for the ici status. We started out using the ici table, but @PaulClark2 suggested using dimcandproperties since there were multiple ways to update ICI and the data might be incomplete. 

I am thinking we might want to strategize on how to do a little extra clean up. 

Let's take Issa for an example:
```
=> select * from dimcandstatusici where cand_sk=10025283;
```

 candstatusici_sk | cand_sk  | election_yr | ici_code | cand_status | cand_inactive_flg |      load_date      | expire_date | pg_date 
------------------|----------|-------------|----------|-------------|-------------------|---------------------|-------------|---------
          1100948 | 10025283 |        2004 |          | C           |                   | 2015-06-24 18:55:17 |             | 
          1073269 | 10025283 |        2008 | I        | C           |                   | 2015-06-24 18:55:17 |             | 
          1073271 | 10025283 |        2012 | I        | P           |                   | 2015-06-24 18:55:17 |             | 
          1100953 | 10025283 |        2016 |          | C           |                   | 2015-06-24 18:55:17 |             | 
          1073268 | 10025283 |        2006 | I        | C           |                   | 2015-06-24 18:55:17 |             | 
          1073270 | 10025283 |        2010 | I        |             |                   | 2015-06-24 18:55:17 |             | 
          1100952 | 10025283 |        2014 |          | C           |                   | 2015-06-24 18:55:17 |             | 


```
c=> select election_yr, cand_ici_cd, cand_status_cd from dimcandproperties where cand_sk=10025283;
```

 election_yr | cand_ici_cd | cand_status_cd 
-------------|-------------|----------------
        2000 | O           | C
        2002 | I           | C
        2004 |             | C
        2004 |             | C
        2004 |             | C
        2006 |             | C
        2008 |             | P
        2008 | I           | P
        2008 | I           | C
        2010 |             | P
        2010 | I           | P
        2010 | I           | P
        2006 | I           | C
        2012 | I           | N
        2012 | I           | N
        2012 | I           | N
        2012 | I           | C
        2014 |             | N
        2014 |             | N
        2014 | I           | N
        2014 | I           | C
        2014 | I           | C
        2016 |             | N
        2016 | I           | N
        2016 | I           | F
        2016 | I           | C

Paul, do you think we could  "fill down" the ici status for each candidate in the dimcandproperties table?

**Comment by LindsayYoung** [comment]
Actually, we might just want to add another variable that is basically-- has this person ever been a statuary candidate? We could then filter out those who haven't from typeahead etc.

File: webservices/args.py (hunk 1 above)
Comment (LindsayYoung) [inline webservices/args.py]: Yea, that is jargon-y I will update.

**Comment by LindsayYoung** [comment]
I changed `year` to `election_year` in candidate detail, because that is what we have on master and I don't want to break people's stuff if we can avoid it.

**Comment by jmcarp** [comment]
@LindsayYoung @PaulClark2 what are your thoughts on this? Do you think this patch would hide many interesting candidates? Does the benefit of hiding irrelevant candidates outweigh the cost of potentially hiding some relevant candidates?

**Comment by LindsayYoung** [comment]
I think this is a good idea, this just gets rid of them in the typeahead, and you can still search for them and they will come up. We might be able to improve it later too.

**Comment by PaulClark2** [comment]
The more I think about this the less I like this idea.   Is the goal just to reduce the number of candidates that users see in the type-ahead?   Or is it to reduce seemingly offensive names popping up.   That happens on our website now.  Have we done any testing to confirm that users are actually confused by seeing more candidates?   @onezerojeremy  Does this cause more problems that it solves?   
Currently using the "statutory candidate" logic will exclude the following presidential candidates:  Lawrence Lessig, Jim Webb, Chris Christie, Jim Gilmore and John Kasich.

Examples from earlier this year (this would happen every cycle):  
 J. Bush filed his F2-2016 on 6/15/15 but would have been missing from the type-ahead until the end of July 2015. 
H. Clinton filed her F2-2016 on 4/13/15 but would have been missing from the type-ahead until the end of July 2015. 
T. Cruz filed his F2-2016 on 3/23/15 but would have been missing from the type-ahead until the end of July 2015.
B. Sanders filed his F2-2016 on 4/30/2015 but would have been missing from the type-ahead until the end of July 2015.

@jwchumley @amykort @amypike thoughts?

**Comment by jwchumley** [comment]
I agree with Paul. It isn't worth losing real candidates just to try to keep offensive names out of the type ahead.

**Comment by LindsayYoung** [comment]
It is a simple thing to roll back.

There was an error where candidates without committees would be eliminated on current fec.gov site and those changes were not being pushed through the data warehouse. Is that still the case, Paul? If so, might we might want to hold off until that is fixed and then go back to the broader search?

**Comment by onezerojeremy** [comment]
Someone remind me what "statutory candidate" means? Sorry. 

We haven't tested specifically for this, but I have heard a smattering of complaints of users looking for "the real" candidates. I know the reality or unreality of the candidates is more complex than the users model, and we haven't tested it in the type-ahead specifically. My gut, which is not data, is that it's a pretty low risk place to experiment with a switch, given that it's a convenience feature (a convenience feature that people LOVE) that doesn't replace the full search. 

Also, it'll be easy to test for confusion: I can add a search for one of those you've mentioned ( Lessig or Webb or whoever) to the testing docket for next week and see if anyone gets tripped up.

**Comment by noahmanger** [comment]
I believe "Statutory candidate" in this sense means someone who has reported $5,000+ raised. 

I think the biggest concern isn't the offensive names, but in posing a barrier to people getting to what they're looking for quickly. Considering that this change would mean that certain legitimate candidates are temporarily not included in the typeahead until they meet the threshold, I see how this could have the contrary effect we're looking for. 

But on the other hand, filing the typeahead with garbage results also poses a barrier. Which means we probably need to make a judgement call about what's better. 

I like the idea of adding a test along those lines.

**Comment by jmcarp** [comment]
This change has been reverted. Unless testing shows that "fake" candidates are a significant barrier to users finding what they're looking for, I agree with Paul and Jeff that we shouldn't hide candidates who don't have status entries. FWIW, @leahbannon ran into this issue and reported a bug almost immediately after we hid the candidates.

An alternative approach is to get back to the question of ordering typeahead search results. They're not shown in any particular order right now, so sorting by funds raised, candidate status, etc. would help bring relevant candidates to the top of the list. This is something @LindsayYoung has brought up before--maybe it's time to implement.

**Comment by onezerojeremy** [comment]
The alternative approach of sorting makes sense to me. 

Confusion for something missing is easier to test for in usability testing than confusion due to signal to noise ratio. Anyone have ideas on what are some particularly noisy searches?

**Comment by LindsayYoung** [comment]
My favorite approach is still weighting by money raised in some way. Perhaps cumulative money raised or how much they raised in their last 2-year cycle. If that is not enough, we might want to add extra weight if a candidate is active.  

Here is an example of the search not doing that great of a job if surfacing the most likely candidates to the top. If you search for Clinton:
 1. Some guy who filed in 1995 and never raised any money
 2. Bill Clinton-- but not President Clinton, A 2016 candidate for the Democratic Farm Labor Party
 3. Hillary Clinton for President
 4. Hillary Clinton for Senate
 5. Former President Bill Clinton

screen shot 2015-10-16 at 10 03 42 am

**Comment by LindsayYoung** [comment]
Here is a committee example. The result people are most likely looking for is listed 5th.

screen shot 2015-10-16 at 10 12 57 am

**Comment by jwchumley** [comment]
what data elements do you have access to for this? Maybe you use max(years active) then name and possibly total receipts?

**Comment by LindsayYoung** [comment]
Right now it is just going off of name last year active should be easy to add.

Here is a previous thread about this too #943 and fuzzy matching came up in #854

**Comment by leahbannon** [comment]
@LindsayYoung I love that idea of weighting by money raised because I think it solves both problems of too much noise and risking hiding candidates too much

**Comment by noahmanger** [comment]
Let's move this conversation over to the original issue for weighting search results: https://github.com/18F/openFEC/issues/943

**Pull request merged**